common/src/main/java/com/linecorp/centraldogma/common/MergeQuery.java

```diff
@@ -0,0 +1,111 @@
+
+import static com.linecorp.centraldogma.common.QueryType.IDENTITY;
+import static com.linecorp.centraldogma.common.QueryType.JSON_PATH;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * A merge query on files.
+ *
+ * @param <T> the content type of files being queried
+ */
+public interface MergeQuery<T> {
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}.
+     *
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     */
+    static MergeQuery<JsonNode> ofJson(MergeSource... mergeSources) {
+        return ofJson(ImmutableList.copyOf(requireNonNull(mergeSources, "mergeSources")));
+    }
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}.
+     *
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     */
+    static MergeQuery<JsonNode> ofJson(Iterable<MergeSource> mergeSources) {
+        return new JsonMergeQuery(IDENTITY, mergeSources, ImmutableList.of());
+    }
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}. Then, the specified
+     * <a href="https://github.com/json-path/JsonPath/blob/master/README.md">JSON path expressions</a>
+     * are applied to the content.
+     *
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     * @param jsonPaths the JSON path expressions to apply
+     */
+    static MergeQuery<JsonNode> ofJsonPath(Iterable<MergeSource> mergeSources,
+                                           String... jsonPaths) {
+        return ofJsonPath(mergeSources, ImmutableList.copyOf(requireNonNull(jsonPaths, "jsonPaths")));
+    }
+
+    static MergeQuery<JsonNode> ofJsonPath(Iterable<MergeSource> mergeSources,
+                                           Iterable<String> jsonPaths) {
+        return new JsonMergeQuery(JSON_PATH, mergeSources, jsonPaths);
+    }
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}. Then, the specified
+     * <a href="https://github.com/json-path/JsonPath/blob/master/README.md">JSON path expressions</a>
+     * are applied to the content.
+     *
+     * @param type the type of the {@link MergeQuery}
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     * @param jsonPaths the JSON path expressions to apply
+     */
+    static MergeQuery<JsonNode> of(QueryType type, Iterable<MergeSource> mergeSources,
+                                   Iterable<String> jsonPaths) {
```

Comment:
Could we make this general like `Query.of()` just in case we introduce merge queries to other formats?